src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs

```diff
@@ -232,5 +284,86 @@ private IEnumerable<LibraryDependency> GetDependencies(FindPackageByIdDependency
                 }
             }
         }
+
+        /// <summary>
+        /// Discover all package versions from a feed.
+        /// </summary>
+        private async Task<IEnumerable<NuGetVersion>> GetPackageVersionsAsync(string id,
+                                                                    SourceCacheContext cacheContext,
+                                                                    ILogger logger,
+                                                                    CancellationToken cancellationToken)
+        {
+            IEnumerable<NuGetVersion> packageVersions = null;
+            try
+            {
+                if (_throttle != null)
+                {
+                    await _throttle.WaitAsync();
+                }
+                packageVersions = await _findPackagesByIdResource.GetAllVersionsAsync(
+                    id,
+                    cacheContext,
+                    logger,
+                    cancellationToken);
+            }
+            catch (FatalProtocolException e) when (_ignoreFailedSources)
+            {
+                if (!_ignoreWarning)
+                {
+                    _logger.LogWarning(e.Message);
+                }
+                return null;
+            }
+            finally
+            {
+                _throttle?.Release();
+            }
+
+            return packageVersions;
+        }
+
+        /// <summary>
+        /// Cache on LibraryRange and TargetFramework
+        /// </summary>
+        private class DependencyInfoCacheKey : IEquatable<DependencyInfoCacheKey>
```

Comment:
As an alternative, this could be just a formatted string as a key. This would eliminate the need in the separate key class.